In [1]:
# Import the necessary packages
import numpy as np
import pandas as pd

In [2]:
# Load the data
data = pd.read_csv("http://rcs.bu.edu/examples/python/data_analysis/Salaries.csv")

# Display the first 3 rows
data.head(3)

rank discipline  phd  service   sex  salary
0  Prof          B   56       49  Male  186960
1  Prof          A   12        6  Male   93000
2  Prof          A   23       20  Male  110515

In [3]:
# Determine the data types of the columns
data.dtypes

rank          object
discipline    object
phd            int64
service        int64
sex           object
salary         int64
dtype: object

In [4]:
# Another method of determining the information of the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   rank        78 non-null     object
 1   discipline  78 non-null     object
 2   phd         78 non-null     int64 
 3   service     78 non-null     int64 
 4   sex         78 non-null     object
 5   salary      78 non-null     int64 
dtypes: int64(3), object(3)
memory usage: 3.8+ KB


In [5]:
# General summary of the numeric variables
data.describe()

phd    service         salary
count  78.000000  78.000000      78.000000
mean   19.705128  15.051282  108023.782051
std    12.498425  12.139768   28293.661022
min     1.000000   0.000000   57800.000000
25%    10.250000   5.250000   88612.500000
50%    18.500000  14.500000  104671.000000
75%    27.750000  20.750000  126774.750000
max    56.000000  51.000000  186960.000000

#### The numerical variables are:
* `phd`
* `service`
* `salary`

In [6]:
# Summary of the categorical variables
data.describe(include='object')

rank discipline   sex
count     78         78    78
unique     3          2     2
top     Prof          B  Male
freq      46         42    39

#### The categorical variables are:
* `rank`
* `discipline`
* `sex`

In [7]:
# Split the variables into numerical and categorical variables
num_df = data.select_dtypes(exclude='O') # exclude object data type
num = num_df.columns.to_list() # extract the column names and convert to a Python list

cat_df = data.select_dtypes(include='O') # include object data type
cat = cat_df.columns.to_list() # extract the column names and convert to a Python list

In [8]:
print(f"Numerical variables:\n {num} \n")
print('--' * 15)
print(f"Categorical variables:\n {cat}")

Numerical variables:
 ['phd', 'service', 'salary'] 

------------------------------
Categorical variables:
 ['rank', 'discipline', 'sex']


In [9]:
# Frequency table of each categorical variable
for col in cat:
    val = data[col].value_counts()
    print(f"\t{col}:")
    print(f"{val} \n")

	rank:
Prof         46
AsstProf     19
AssocProf    13
Name: rank, dtype: int64 

	discipline:
B    42
A    36
Name: discipline, dtype: int64 

	sex:
Male      39
Female    39
Name: sex, dtype: int64 



#### Comment:
* There are more __Professors__ (rank).
* The category labelled **B** is greater than **A** (discipline).
* The number of **males** and **females** are equal (sex).

In [10]:
data.head()

rank discipline  phd  service   sex  salary
0  Prof          B   56       49  Male  186960
1  Prof          A   12        6  Male   93000
2  Prof          A   23       20  Male  110515
3  Prof          A   40       31  Male  131205
4  Prof          B   20       18  Male  104800

#### Descriptive Statistics

In [11]:
data[num].describe()

phd    service         salary
count  78.000000  78.000000      78.000000
mean   19.705128  15.051282  108023.782051
std    12.498425  12.139768   28293.661022
min     1.000000   0.000000   57800.000000
25%    10.250000   5.250000   88612.500000
50%    18.500000  14.500000  104671.000000
75%    27.750000  20.750000  126774.750000
max    56.000000  51.000000  186960.000000

#### Comment:
+ **PhD:** mean = 19.7, median = 18.5, max(100th percentile) = 56, std = 12.5
+ **Service:** mean = 15.1, median = 14.5, max(100th percentile) = 51, std = 12.1
+ **PhD:** mean = 108024, median = 104671, max(100th percentile) = 186960, std = 28293.7

### Aggregations

<br>

#### Which rank earned the most?

In [12]:
# Which rank earned the most? i.e earnings by rank
grp_rank = data.groupby('rank')
grp_rank[['salary']].agg(['mean', 'median'])

salary          
                    mean    median
rank                              
AssocProf   91786.230769  103613.0
AsstProf    81362.789474   78500.0
Prof       123624.804348  123321.5

In [13]:
# Earnings in descending order
grp_rank[['salary']].agg(['mean', 'median']).apply(lambda x: x.sort_values(ascending=False))

salary          
                    mean    median
rank                              
Prof       123624.804348  123321.5
AssocProf   91786.230769  103613.0
AsstProf    81362.789474   78500.0

#### Which rank had the most PhD?

**Note:** I used 'median' in my computation because it's not affected by outliers unlike 'mean'.

In [14]:
grp_rank[['phd']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

phd
          median
rank            
Prof        24.5
AssocProf   13.0
AsstProf     4.0

#### Which rank had the most 'service'?

In [15]:
grp_rank[['service']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

service
           median
rank             
Prof           19
AssocProf       9
AsstProf        2

#### Which gender earned the most?

In [16]:
data.groupby('sex')[['salary']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

salary
        median
sex           
Male    107300
Female  103750

#### Which gender has more PhDs?

In [17]:
data.groupby('sex')[['phd']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

phd
       median
sex          
Male       21
Female     17

#### Which 'gender' had more 'service'?

In [18]:
data.groupby('sex')[['service']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

service
        median
sex           
Male        19
Female      10

#### Which discipline earned the most?

In [19]:
data.groupby('discipline')[['salary']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

salary
            median
discipline        
B           110581
A            93675

#### Which discipline had the most PhDs?

In [20]:
data.groupby('discipline')[['phd']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

phd
           median
discipline       
A            21.5
B            17.0

#### Which discipline had the most service?

In [21]:
data.groupby('discipline')[['service']].agg(['median']).apply(lambda x: x.sort_values(ascending=False))

service
            median
discipline        
A             16.0
B             12.5

#### Who earned the most?

In [22]:
# Max salary
max_salary = data['salary'].max()

data[data['salary'] == max_salary]

rank discipline  phd  service   sex  salary
0  Prof          B   56       49  Male  186960

#### Who earned the least?

In [23]:
# Min salary
min_salary = data['salary'].min()

data[data['salary'] == min_salary]

rank discipline  phd  service   sex  salary
9  Prof          A   51       51  Male   57800

#### Which female earned the most?

In [24]:
females = data[data['sex'] == 'Female'] # dataframe containing only females
max_fem_salary = females['salary'].max() # max salary

# highest earning female
females[females['salary'] == max_fem_salary]

rank discipline  phd  service     sex  salary
72  Prof          B   24       15  Female  161101

#### Which female earned the least?

In [25]:
min_fem_salary = females['salary'].min() # min salary

# least earning female
females[females['salary'] == min_fem_salary]

rank discipline  phd  service     sex  salary
54  AssocProf          A   25       22  Female   62884